### ISSUE TITLE
Removal of inventory windows happens after player data saving

### ISSUE THREAD

**Issue by UnknownOre** [pmmp/PocketMine-MP#4129]
### Issue description

- Expected result: player data saving after all calls involves the player is executed
- Actual result: player data is being saved too early

### Steps to reproduce the issue
1. use the code below.
2. open a chest
3. kick the player
4. relog

### OS and versions

* PocketMine-MP: 3.18.0
* PHP: 7.4
* Server OS: Windows 10

### Plugins
```php
<?php
namespace Test;

use pocketmine\block\Chest;
use pocketmine\event\Listener;
use pocketmine\event\player\PlayerInteractEvent;
use pocketmine\plugin\PluginBase;

class Loader extends PluginBase implements Listener {


    public function onEnable(){
        $this->getServer()->getPluginManager()->registerEvents($this,$this);
    }

    public function onInteract(PlayerInteractEvent $event){
        if($event->getBlock() instanceof Chest){
            $tile = $event->getBlock()->getLevel()->getTile($event->getBlock());
            if($tile instanceof \pocketmine\tile\Chest){
                $event->setCancelled();
                $event->getPlayer()->addWindow(new DoubleChest($tile));
            }
        }
    }
}

use pocketmine\inventory\ChestInventory;
use pocketmine\item\Item;
use pocketmine\Player;

class DoubleChest extends ChestInventory
{


    public function onClose(Player $who): void
    {
        parent::onClose($who);
        var_dump("onClose was run");
        $who->getInventory()->addItem(Item::get(Item::DIRT));
    }

}
```

**Comment by dktapps** [comment]
I'm not clear what the issue is here. Please be more specific.

**Comment by Dapro718** [comment]
@UnknownOre after the chest is closed, can the player open another chest or their own inventory?

**Comment by UnknownOre** [comment]
> I'm not clear what the issue is here. Please be more specific.

using the code above you will not find the dirt item inside your inventory after relogging, because pocketmine saved the player data before that item got added.

**Comment by dktapps** [comment]
I guess this issue is caused by this running after the save data handler fires. https://github.com/pmmp/PocketMine-MP/blob/stable/src/pocketmine/Player.php#L3737

**Comment by dktapps** [comment]
This should have been fixed in PM4 by c7beb0a70258fdf5dbba6bf4f93b17cca2a88838.
